### ISSUE TITLE
What about limits

### ISSUE THREAD

**Issue by smalot** [containous/traefik#2016]
Hi,

I wonder if there is a limit for frontends and backends for a single installation.
My goal is to organize something around 400 backend containers and more than 100'000 frontend domain names (may be more).
Can I do such thing with Traefik ?
Do I need specific hardware (memory / cpu) to do the job ?

I can't find any limit page in the documentation

Many thanks

**Comment by ldez** [comment]
Thanks for your interest in Traefik !

Every question in the issues means features less 😉.

The issue tracker is for reporting bugs and feature requests only.
For end-user related support questions, refer to one of the following:

- [Stack Overflow](https://stackoverflow.com/questions/tagged/traefik) using the "traefik" tag
- the [Traefik community Slack channel](https://traefik.herokuapp.com)

**Comment by ldez** [comment]
It is very difficult to define limits:
- the project changes very quickly
- it depends on qps and the size of queries
- it depends on middlewares (auth, redirect, retry, ...)
- it depends on the providers (Docker, k8s, Mesos, ...) and the number of providers that you use
- it depends on the number of backends and frontends
- it depends on the hardware
- benchmarks rarely give realistic data and are very difficult to maintain

I hope you better understand the absence of a page on limitations.

FYI: The previous answer and labels were set by a bot using one of my GH tokens. (now the bot has this own GH token)

**Comment by kajmagnus** [comment]
@smalot  I'm in a similiar situation as you: I'll have maybe 1 000 or 10 000 frontend hostnames (only one or maybe a few backends though).

How did things work out for you? Did you run into any limits? How many frontend hostnames do you have currently? (if I may ask)  Did you re-post this question elsewhere? (then I'd like to go there and have a look :-)  )

@ldez  I'm interpreting what you wrote as that there aren't any known restrictions (say, a hardcoded list that can grow to at most 1 000 items). Instead, if the server load caused by each separate frontend / hostname, is low (in comparison to the server's hardware), then, I can have really many frontends and hostnames.

**Comment by kajmagnus** [comment]
(I had posted a question at SO already before I found this topic here in GitHub:  https://stackoverflow.com/questions/53994690/any-number-of-hostnames-limit-for-traefiks-rule-hosthostname1-name2  )

**Comment by marker** [comment]
Did you find something to workaround traefik's limits?  It's very easy to test and prove that traefik has a very hard-set max limit on frontends/backends.  Simply start traefik configured to dynamically add frontends and backends based on container labels, then start spinning up containers behind it.  The first 115 or so will work perfectly, you'll see both frontends and backends created and listed in the dashboard.  But you'll soon hit a magic limit where all containers added will 404 (just using the basic nginx container), no frontend gets created and no backend either.  The services are all present and running in the swarm, but traefik is overwhelmed and can't create any more entries.

If you then restart traefik, you find it's completely unable to start.  It will hang forever, swarm will show its desired state as "running" but its current state is stuck at "New", where it will remain for 15+ minutes until you give up and kill the whole stack.  You can get it to start by pruning your nginx containers, but it will die again when you re-add them.

TLDR: A very easy test proves that traefik definitely has hard and fast limits of around 115-120 frontend/backend pairs and going beyond that kills it entirely.

So, what did you commenters above end up using?  We're needing to reevaluate load balancer choices since traefik definitely can't handle nearly 1,200 frontends each with a distinct backend.

**Comment by emilevauge** [comment]
@marker Thank you for your comment 😃 

I don't know where to start...

Of course, there is **no** limitation hard-coded nor configured intp Traefik on the maximum frontends/backends it can handle. I don't see any good reason why we would have done this. The code can easily be reviewed by anyone to check this, as it is open sourced. And yes, it is reviewed by many people all the time.

We did some benchmark tests and we have been able to route to more than 100k frontends.

Hitting limitations does not imply that things are hard coded somewhere. For example, you can reach limits because you deployed more containers on your server it can handle. Or maybe your issue is due to a real bug with our Swarm provider implementation! One thing is sure, it is obviously too early to draw any conclusions on your issue. I invite you to contact us on Slack and give more details on your configuration.
